collectors/osidb_scheduler/collectors.py

```diff
@@ -0,0 +1,42 @@
+"""
+OSIDB Collectors
+"""
+from celery.utils.log import get_task_logger
+
+from collectors.framework.models import Collector
+from osidb.mixins import Alert
+
+logger = get_task_logger(__name__)
+
+
+class StaleAlertCollector(Collector):
+    """
+    Stale Alert Collector
+    """
+
+    def collect(self):
+        """
+        collector run handler
+
+        On every run, this collector will check if the alert is
+        still valid by comparing the creation time of the alert
+        with the validation time of the Model.
+
+        If the creation time of the alert is older than
+        the validation time of the Model,
+        the alert is considered stale and will be deleted.
+        """
+        logger.info("Collecting Stale Alerts")
+
+        alerts = Alert.objects.all()
```

Comment:
I'm suspecting this will take a long time to finish, we current have ~788k alerts in stage and querying them with all related objects would make this collector interact in almost all entries of the database, I suggest trying querying only for duplicated entries with something like
```
duplicate_alerts = Alert.objects.values('object_id', 'content_object').annotate(
    last_validated_dt_count=Count('last_validated_dt', distinct=True)
).filter(last_validated_dt_count__gt=1)
```

Reply:
Hmm yeah I need to improve my Django filters knowledge 😂 
I'll do some testing to improve it, thanks for the suggestion!

Reply:
I have updated the collector with a more in depth logic to filter the alerts directly on the database level using subqueries.

I'm nowhere near the 788k alerts in the database, but here are the logs with 45k alerts, it takes 0.9 seconds to complete:
```
2024-12-11 11:34:14,427 [INFO] collectors.osidb_scheduler.collectors: Searching for stale alerts in 2 content types
2024-12-11 11:34:14,951 [INFO] collectors.osidb_scheduler.collectors: Found 1882 stale alerts
2024-12-11 11:34:15,233 [INFO] collectors.osidb_scheduler.collectors: Deleted 1882 stale alerts
```
I'll continue to run tests as the database grows, to see the performance

Reply:
Thank you for addressing it!